numpy/_core/numeric.py

```diff
@@ -1121,6 +1121,72 @@ def tensordot(a, b, axes=2):
     return res.reshape(olda + oldb)
 
 
+def _vecdot_dispatcher(x1, x2, /, *, axis=None):
+    return (x1, x2)
+
+
+@array_function_dispatch(_vecdot_dispatcher)
+def vecdot(x1, x2, /, *, axis=-1):
+    """
+    Computes the (vector) dot product of two arrays.
+
+    Let :math:`\\mathbf{a}` be a vector in `x1` and :math:`\\mathbf{b}` be
+    a corresponding vector in `x2`. The dot product is defined as:
+
+    .. math::
+       \\mathbf{a} \\cdot \\mathbf{b} = \\sum_{i=0}^{n-1} \\overline{a_i}b_i
+
+    over the dimension specified by `axis` and where :math:`n` is
+    the dimension size and :math:`\\overline{a_i}` denotes the complex
+    conjugate if :math:`a_i` is complex and the identity if :math:`a_i`
+    is real-valued.
+
+    .. note::
+
+        This function differs from :py:func:`np.vdot <numpy.vdot>` as
+        :py:func:`np.vdot <numpy.vdot>` additionally flattens input arguments.
+
+    This function is Array API compatible.
+
+    Parameters
+    ----------
+    x1 : array_like
+        First input array.
+    x2 : array_like
+        Second input array.
+    axis : int, optional
+        Axis over which to compute the dot product. Default: ``-1``.
+
+    Returns
+    -------
+    output : ndarray
+        The vector dot product of the input.
+
+    See Also
+    --------
+    dot
+
+    """
+    ndim = builtins.max(x1.ndim, x2.ndim)
+    x1_shape = (1,)*(ndim - x1.ndim) + tuple(x1.shape)
+    x2_shape = (1,)*(ndim - x2.ndim) + tuple(x2.shape)
+    if x1_shape[axis] != x2_shape[axis]:
+        raise ValueError(
+            "x1 and x2 must have the same size along the dot-compute axis "
+            f"but they are: {x1_shape[axis]} and {x2_shape[axis]}."
+        )
+
+    x1_, x2_ = np.broadcast_arrays(x1, x2)
```

Comment:
Darn, another one. Would it still be possible to add `subok=True` here to allow subclasses to "just work"?